### ISSUE TITLE
Optimistic compression approach

### ISSUE THREAD

**Issue by kimstik** [YosysHQ/apicula#235]
The following code of write_bitstream() will hang on data that has less than three unused values:
```python
        lst = bitmatrix.histogram(bs, bins=[i for i in range(257)]) # 257 iso that the last basket is [255, 256] and not [254, 255]
        [key8Z, key4Z, key2Z] = [i for i,val in enumerate(lst) if val==0][0:3]
```

**Comment by yrabbit** [comment]
Can you suggest a fix? It's just not my area at all.

**Comment by trabucayre** [comment]
Indeed, this code may produce error/crash if less than 3 keys are found.
Have you a design with this situation occur? Is it possible to you share it? I'm interested to see how gowin deal with this use case (no compression or compression with 1 or 2 keys instead of 3).
Quick fix may be to check the size and disable compression when len < 3.

**Comment by kimstik** [comment]
I'll need a bit of time for PR. The quick fix doesn't work because it requires to turn off compression later.

**Comment by yrabbit** [comment]
Now we will do the following: as many keys as there are unused bytes, so many will be used - if there is only one unused byte, then there will be only one key.

In a completely bleak case when all bytes are unused, compression will be turned off, and the header and pad will be changed accordingly.

https://github.com/YosysHQ/apicula/pull/324/commits/0ac9be786e51731f9c26ba23c977ae71c03ab68a
